api/spec/v1.yaml

```diff
@@ -1614,6 +1675,76 @@ components:
       description: |
         A ParaTime block.
 
+    RuntimeEventList:
+      type: object
+      required: [events]
+      properties:
+        events:
+          type: array
+          items:
+            $ref: '#/components/schemas/RuntimeEvent'
+      description: |
+        A list of runtime events.
+
+    RuntimeEvent:
+      type: object
+      required: [round, type, body]
+      properties:
+        round:
+          type: integer
+          format: int64
+          description: The block height at which this event was generated.
+          example: *runtime_block_round_1
+        tx_hash:
+          type: string
+          nullable: true
+          description: |
+            Hash of this event's originating transaction
+            Absent if the event did not originate from a transaction.
+          example: *tx_hash_1
+        type: 
+          description: The type of the event.
+          $ref: '#/components/schemas/RuntimeEventType'
+        body:
+          type: object
+          description: |
+            The decoded event contents. This spec does not encode the many possible types;
+            instead, see [the Go API](https://pkg.go.dev/github.com/oasisprotocol/oasis-sdk/client-sdk/go/modules).
+            This object will conform to one of the `*Event` types two levels down
+            the hierarchy, e.g. `MintEvent` from `Event > accounts.Event > MintEvent`,
+            OR one of the known runtime event types enumerated in EvmRuntimeEvent.
+      description: An event emitted by the runtime layer
+
+    RuntimeEventType:
+      type: string
+      enum:
+        - accounts.transfer
+        - accounts.burn
+        - accounts.mint
+        - consensus_accounts.deposit
+        - consensus_accounts.withdraw
+        - core.gas_used
+        - evm.erc20_transfer
+      example: *runtime_event_type_1
+
+    EvmRuntimeEvent:
+      type: object
+      enum:
+        - properties:
+            from:
+              type: string
+              description: The evm address the value was transferred from.
+              example: *runtime_address_1
+            to:
+              type: string
+              description: The evm address the value was transferred to.
+              example: *runtime_address_2
+            value:
+              type: uint64
+              description: The balance that was transferred. (in wei? nano-ROSE?)
+              example: 1000000000
+          description: An ERC20 Transfer Event.
```

Comment:
I _don't_ think we can have object enums; but leaving this here as a discussion starting point

Reply:
You mean "sum types", aka [lots of other names](https://en.wikipedia.org/wiki/Tagged_union)?
openapi has `oneOf`, though I don't expect it maps elegantly to Go (which doesn't have native support for sum types, unlike e.g. Rust or Typescript)

Regardless, we don't want to enumerate the possible types here. Right now we only support erc20 Transfer events. But eventually, we'll have "contract verification", meaning that we'll be able to pull the ABIs of most (hopefully most, anyway) contracts, meaning we'll be able to "understand"/parse those contracts' events, and the result will be of an arbitrary type. It won't (necessarily) have any of the `from`, `to`, `value` fields. 

FWIW, right now, we're able to understand/parse erc20 transfer events because we [hardcoded](https://github.com/oasisprotocol/oasis-indexer/blob/1016ed82c9010370291719a13622a923f9d403a0/analyzer/evmabi/contracts/artifacts/ERC20.json#L230-L251) its ABI into our source code.

So the whole `EvmRuntimeEvent` type is too specific; we won't use it, or anything like it.

Reply:
Neat, I didn't know about tagged unions before.

> we'll be able to "understand"/parse those contracts' events, and the result will be of an arbitrary type

Just curious - are we expecting to have some automation for the "contract verification," so parsing other runtimeEvent types won't require manual parsing / hardcoding?

> So the whole EvmRuntimeEvent type is too specific

Got it, will remove